### PR TITLE
[ENH] Add A MultiSurfaceMasker class 

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Relates to #5539 

Changes proposed in this pull request:

- moves several methods from the multi maskers into the MultiMixin
- create a MultiSurfaceMasker
- changelog
- adapt tests

### CODE REVIEW

File: nilearn/maskers/multi_nifti_masker.py [hunk 1]
```diff
@@ -441,15 +438,9 @@ def transform_imgs(
...
-        confounds = prepare_confounds_multimaskers(self, imgs_list, confounds)
+        confounds = self._prepare_confounds(imgs_list, confounds)
 
-        if sample_mask is None:
-            sample_mask = itertools.repeat(None, len(imgs_list))
-        elif len(sample_mask) != len(imgs_list):
-            raise ValueError(
-                f"number of sample_mask ({len(sample_mask)}) unequal to "
-                f"number of images ({len(imgs_list)})."
-            )
```
Comment (Remi-Gau) [inline nilearn/maskers/multi_nifti_masker.py:449]: extracted in MutliMixin

File: nilearn/maskers/multi_nifti_masker.py [hunk 2]
```diff
@@ -493,52 +484,3 @@ def transform_imgs(
             )
         )
         return data
-
-    @fill_doc
-    def transform(self, imgs, confounds=None, sample_mask=None):
```
Comment (Remi-Gau) [inline nilearn/maskers/multi_nifti_masker.py:495]: extracted in MutliMixin

File: nilearn/maskers/multi_nifti_maps_masker.py [hunk 3]
```diff
@@ -182,106 +174,3 @@ def __init__(
             cmap=cmap,
             clean_args=clean_args,
         )
-
-    @fill_doc
-    def transform_imgs(
```
Comment (Remi-Gau) [inline nilearn/maskers/multi_nifti_maps_masker.py:187]: extracted in MutliMixin

File: nilearn/maskers/multi_nifti_maps_masker.py [hunk 4]
```diff
@@ -182,106 +174,3 @@ def __init__(
...
-        region_signals = Parallel(n_jobs=n_jobs)(
-            delayed(func)(imgs=imgs, confounds=cfs, sample_mask=sms)
-            for imgs, cfs, sms in zip(
-                niimg_iter, confounds, sample_mask, strict=False
-            )
-        )
-        return region_signals
-
-    @fill_doc
-    def transform(self, imgs, confounds=None, sample_mask=None):
```
Comment (Remi-Gau) [inline nilearn/maskers/multi_nifti_maps_masker.py:243]: extracted in MutliMixin

File: nilearn/maskers/multi_nifti_labels_masker.py [hunk 5]
```diff
@@ -179,105 +171,3 @@ def __init__(
             clean_args=clean_args,
             keep_masked_labels=keep_masked_labels,
         )
-
-    @fill_doc
-    def transform_imgs(
```
Comment (Remi-Gau) [inline nilearn/maskers/multi_nifti_labels_masker.py:184]: extracted in MutliMixin

File: nilearn/maskers/multi_nifti_labels_masker.py [hunk 6]
```diff
@@ -179,105 +171,3 @@ def __init__(
...
-        region_signals = Parallel(n_jobs=n_jobs)(
-            delayed(func)(imgs=imgs, confounds=cfs, sample_mask=sms)
-            for imgs, cfs, sms in zip(
-                niimg_iter, confounds, sample_mask, strict=False
-            )
-        )
-        return region_signals
-
-    @fill_doc
-    def transform(self, imgs, confounds=None, sample_mask=None):
```
Comment (Remi-Gau) [inline nilearn/maskers/multi_nifti_labels_masker.py:239]: extracted in MutliMixin

File: nilearn/maskers/base_masker.py [hunk 7]
```diff
@@ -182,42 +180,6 @@ def filter_and_extract(
     return region_signals, aux
 
 
-def prepare_confounds_multimaskers(masker, imgs_list, confounds):
```
Comment (Remi-Gau) [inline nilearn/maskers/base_masker.py:178]: extracted in MutliMixin

File: nilearn/decomposition/_base.py [hunk 8]
```diff
@@ -512,11 +520,22 @@ def fit(self, imgs, y=None, confounds=None):
 
         self._fit_cache()
 
+        if self.mask is not None:
+            assert isinstance(
```
Comment (Remi-Gau) [inline nilearn/decomposition/_base.py]: TODO: throw a proper error

File: nilearn/decomposition/_multi_pca.py [hunk 9]
```diff
@@ -32,10 +32,10 @@ class _MultiPCA(_BaseDecomposition):
 
     %(smoothing_fwhm)s
 
-    mask : Niimg-like object, :obj:`~nilearn.maskers.NiftiMasker` or \
-          :obj:`~nilearn.maskers.MultiNiftiMasker` or \
+    mask : Niimg-like object,  :obj:`~nilearn.maskers.MultiNiftiMasker` or \
            :obj:`~nilearn.surface.SurfaceImage` or \
-           :obj:`~nilearn.maskers.SurfaceMasker` object, optional
```
Comment (Remi-Gau) [inline nilearn/decomposition/_multi_pca.py:38]: TODO the common doc string should probably be refactored

File: nilearn/regions/tests/test_parcellations.py [hunk 10]
```diff
@@ -347,7 +347,7 @@ def test_parcellation_all_methods_with_surface(method, n_parcels, rng):
         ),
     }
     surf_img = SurfaceImage(mesh=mesh, data=data)
-    parcellate = Parcellations(method=method, n_parcels=n_parcels)
```
Comment (Remi-Gau) [inline nilearn/regions/tests/test_parcellations.py:350]: unrelated to makes the tests output less verbose

File: nilearn/maskers/tests/test_surface_masker.py [hunk 11]
```diff
@@ -56,24 +56,24 @@ def test_check_estimator_nilearn(estimator, check, name):  # noqa: ARG001
...
 
 
-def test_fit_list_surf_images_with_mask(surf_mask_1d, surf_img_2d):
+def test_fit_list_surf_images_with_mask(surf_mask_1d, surf_img_1d):
     """Test fit on list of surface images when masker has a mask."""
     masker = SurfaceMasker(mask_img=surf_mask_1d)
-    masker.fit([surf_img_2d(3), surf_img_2d(5)])
-    assert masker.mask_img_.shape == (surf_img_2d(1).shape[0],)
+    masker.fit([surf_img_1d, surf_img_1d])
+    assert masker.mask_img_.shape == (surf_img_1d.shape[0],)
```
Comment (Remi-Gau) [inline nilearn/maskers/tests/test_surface_masker.py]: TODO: check if equivalent tests do not already exist in estimator_check

Comment (Remi-Gau) [reply nilearn/maskers/tests/test_surface_masker.py]: there are equivalent tests: removing

File: nilearn/decomposition/_multi_pca.py (hunk 9 above)
Comment (Remi-Gau) [reply nilearn/decomposition/_multi_pca.py:38]: done

File: nilearn/maskers/surface_labels_masker.py [hunk 12]
```diff
@@ -294,6 +294,16 @@ def fit(self, imgs=None, y=None):
...
 
+            if isinstance(imgs, SurfaceImage) and any(
+                hemi.ndim > 2 for hemi in imgs.data.parts.values()
+            ):
+                raise ValueError(
+                    "should only be SurfaceImage should 1D or 2D."
+                )
+            elif hasattr(imgs, "__iter__"):
+                for i, x in enumerate(imgs):
+                    x.data._check_n_samples(1, f"imgs[{i}]")
```
Comment (Remi-Gau) [inline nilearn/maskers/surface_labels_masker.py:236]: TODO in a follow up PR: refactor this across surface maskers

File: nilearn/maskers/tests/test_surface_masker.py [hunk 13]
```diff
@@ -56,26 +56,6 @@ def test_check_estimator_nilearn(estimator, check, name):  # noqa: ARG001
...
-    masker.fit([surf_img_2d(3), surf_img_2d(5)])
-    assert masker.mask_img_.shape == (surf_img_2d(1).shape[0],)
-    assert masker.mask_img_.shape == (masker.n_elements_,)
-
-
-def test_fit_list_surf_images_with_mask(surf_mask_1d, surf_img_2d):
-    """Test fit on list of surface images when masker has a mask."""
-    masker = SurfaceMasker(mask_img=surf_mask_1d)
-    masker.fit([surf_img_2d(3), surf_img_2d(5)])
-    assert masker.mask_img_.shape == (surf_img_2d(1).shape[0],)
```
Comment (Remi-Gau) [inline nilearn/maskers/tests/test_surface_masker.py:76]: replaced by other tests in nilearn/_utils/estimator_checks

**Comment by Remi-Gau** [comment]
@man-shu @bthirion @hndgzkn 
this PR is big: do you want me to split it in 2?
- one with the things that were extracted into a _MultiMixin
- the other with the new functionality

File: doc/manipulating_images/masker_objects.rst [hunk 14]
```diff
@@ -524,11 +524,23 @@ but to extract data from :class:`~nilearn.surface.SurfaceImage`.
...
-as well as list of 1D or 2D surface data with the same underlying mesh.
+or list of 1D or 2D surface data with the same underlying mesh.
 
 .. tip::
     Surface masker objects can transform both 1D, 2D,
+    as well as list of 1D surface image objects.
+
+    - transforming a 1D image (n_vertices,) produces a 1D array,
+    - transforming a 2D image (n_vertices, samples) produces a 2D array,
+    - transforming a a list of ``length==n`` of 1D image (n_vertices,) produces a 2D array (n_vertices, n)
```
Comment (hndgzkn) [inline doc/manipulating_images/masker_objects.rst]:
```suggestion
    - transforming a list of ``length==n`` of 1D image (n_vertices,) produces a 2D array (n_vertices, n)
```

File: doc/manipulating_images/masker_objects.rst [hunk 15]
```diff
@@ -524,11 +524,23 @@ but to extract data from :class:`~nilearn.surface.SurfaceImage`.
...
+
+    Multi surface masker objects can transform both 1D, 2D,
     as well as list of 1D or 2D surface image objects.
+
+    - transforming a 1D image (n_vertices,) produces a 1D array,
+    - transforming a 2D image (n_vertices, samples) produces a 2D array,
+    - transforming a list of ``length==n`` of images (n_vertices,) produces a list of ``length==n`` of arrays where the dimension of each array matches that of the input image
```
Comment (hndgzkn) [inline doc/manipulating_images/masker_objects.rst:542]: It might be better to be more explicit like above. As far as I understand, it is possible to transform a list of 2D images as well. However `transforming a list of ``length==n`` of  images (n_vertices,)` is not correct for 2D images, or am I mistaken?

File: nilearn/_utils/masker_validation.py [hunk 16]
```diff
@@ -50,10 +50,17 @@ def check_embedded_masker(estimator, masker_type="multi_nii", ignore=None):
...
+    from nilearn.maskers import (
+        MultiNiftiMasker,
+        MultiSurfaceMasker,
+        NiftiMasker,
+        SurfaceMasker,
+    )
 
     if masker_type == "surface":
         masker_type = SurfaceMasker
+    elif masker_type == "multi_surface":
```
Comment (hndgzkn) [inline nilearn/_utils/masker_validation.py:63]: `multi_surface` is missing as `masker_type` in docstring.

File: nilearn/_utils/masker_validation.py [hunk 17]
```diff
@@ -50,10 +50,17 @@ def check_embedded_masker(estimator, masker_type="multi_nii", ignore=None):
         New masker
 
     """
-    from nilearn.maskers import MultiNiftiMasker, NiftiMasker, SurfaceMasker
+    from nilearn.maskers import (
+        MultiNiftiMasker,
+        MultiSurfaceMasker,
+        NiftiMasker,
+        SurfaceMasker,
+    )
```
Comment (hndgzkn) [inline nilearn/_utils/masker_validation.py:59]: Could `check_embedded_masker` function be moved to `nilearn.maskers`?  It seems like it belongs there more. It would help avoid this import inside the function as well.

It is imported only by `glm`, `decoders` and `decomposition` packages which already have dependency on `nilearn.maskers`.

Comment (hndgzkn) [reply nilearn/_utils/masker_validation.py:59]: May be the whole module `masker_validation` can be moved?

File: nilearn/_utils/niimg_conversions.py [hunk 18]
```diff
@@ -152,65 +155,73 @@ def iter_check_niimg(
...
-                        "reference FOV.\n"
-                        f"Reference affine:\n{ref_fov[0]!r}\n"
-                        f"Image affine:\n{niimg.affine!r}\n"
-                        f"Reference shape:\n{ref_fov[1]!r}\n"
-                        f"Image shape:\n{niimg.shape!r}\n"
-                    )
-                from nilearn.image import (
-                    resample_img,  # we avoid a circular import
+        if isinstance(niimg, SurfaceImage):
+            # TODO do some checks
```
Comment (bthirion) [inline nilearn/_utils/niimg_conversions.py:162]: Do you want to do this in another PR ?

File: nilearn/maskers/surface_labels_masker.py [hunk 19]
```diff
@@ -436,6 +446,14 @@ def transform_single_imgs(self, imgs, confounds=None, sample_mask=None):
         check_compatibility_mask_and_images(self.labels_img_, imgs)
         check_polymesh_equal(self.labels_img_.mesh, imgs.mesh)
 
+        if isinstance(imgs, SurfaceImage) and any(
+            hemi.ndim > 2 for hemi in imgs.data.parts.values()
+        ):
+            raise ValueError("should only be SurfaceImage should 1D or 2D.")
```
Comment (bthirion) [inline nilearn/maskers/surface_labels_masker.py]:
```suggestion
            raise ValueError("SurfaceImage should only be 1D or 2D.")
```

File: nilearn/_utils/estimator_checks.py [hunk 20]
```diff
@@ -2764,37 +2738,222 @@ def check_surface_masker_fit_with_mask(estimator):
...
+                assert img.shape == (_make_surface_img().mesh.n_vertices, 1)
+
+            elif n_sample == 0:
+                assert signals.size == estimator.n_elements_
+
+                if isinstance(estimator, (SurfaceLabelsMasker)):
+                    # TODO having a test where SurfaceLabelsMasker
+                    # with mask leads to have only 1 ROI
+                    # is a bit of an edge case
+                    # Should this be changed?
```
Comment (bthirion) [inline nilearn/_utils/estimator_checks.py:2907]: No rush I would say

File: nilearn/_utils/niimg_conversions.py (hunk 18 above)
Comment (Remi-Gau) [reply nilearn/_utils/niimg_conversions.py:162]: yup this one has grown bigger than me after a xmas feast

File: nilearn/maskers/surface_masker.py [hunk 21]
```diff
@@ -220,6 +220,20 @@ def fit(self, imgs=None, y=None):
...
 
+            if isinstance(imgs, SurfaceImage) and any(
+                hemi.ndim > 2 for hemi in imgs.data.parts.values()
+            ):
+                raise ValueError(
+                    "should only be SurfaceImage should 1D or 2D."
+                )
+            elif hasattr(imgs, "__iter__"):
+                for i, x in enumerate(imgs):
+                    x.data._check_n_samples(1, f"imgs[{i}]")
```
Comment (hndgzkn) [inline nilearn/maskers/surface_masker.py:231]: This check seems to repeat above as well. Would it be possible to extract and reuse it where necessary?

File: nilearn/maskers/surface_masker.py [hunk 22]
```diff
@@ -284,9 +298,16 @@ def transform_single_imgs(
...
         check_polymesh_equal(self.mask_img_.mesh, imgs.mesh)
 
+        if isinstance(imgs, SurfaceImage) and any(
+            hemi.ndim > 2 for hemi in imgs.data.parts.values()
+        ):
+            raise ValueError("should only be SurfaceImage should 1D or 2D.")
+        elif hasattr(imgs, "__iter__"):
+            for i, x in enumerate(imgs):
+                x.data._check_n_samples(1, f"imgs[{i}]")
+
```
Comment (hndgzkn) [inline nilearn/maskers/surface_masker.py:310]: Same here.

File: nilearn/maskers/surface_masker.py (hunk 21 above)
Comment (Remi-Gau) [reply nilearn/maskers/surface_masker.py:231]: yup: see my comment https://github.com/nilearn/nilearn/pull/5679#discussion_r2367877913

File: nilearn/_utils/masker_validation.py (hunk 17 above)
Comment (Remi-Gau) [reply nilearn/_utils/masker_validation.py:59]: yeah I think this should happen, but not in this PR

File: doc/manipulating_images/masker_objects.rst (hunk 15 above)
Comment (Remi-Gau) [reply doc/manipulating_images/masker_objects.rst:542]: yeah that entire page will need an update given all the new things about surface data

File: nilearn/maskers/surface_masker.py (hunk 21 above)
Comment (hndgzkn) [reply nilearn/maskers/surface_masker.py:231]: Sorry, I've missed it :/

Comment (Remi-Gau) [reply nilearn/maskers/surface_masker.py:231]: no worries 😉

File: nilearn/_utils/masker_validation.py (hunk 17 above)
Comment (Remi-Gau) [reply nilearn/_utils/masker_validation.py:59]: > May be the whole module `masker_validation` can be moved?

actually not possible because of some circular import with nilearn.image but I tried to move out some things in #5724
